libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Generic.cpp

```diff
@@ -107,60 +89,68 @@ void AP_BattMonitor_SMBus_Generic::timer()
         _state.last_time_micros = tnow;
     }
 
-    read_full_charge_capacity();
-
     // FIXME: Perform current integration if the remaining capacity can't be requested
     read_remaining_capacity();
 
     read_temp();
-
-    read_serial_number();
-
-    read_cycle_count();
 }
 
-// read_block - returns number of characters read if successful, zero if unsuccessful
-uint8_t AP_BattMonitor_SMBus_Generic::read_block(uint8_t reg, uint8_t* data, bool append_zero) const
+// Read battery information messages  PEC, cycles, serial number, device name, manufacturer name, etc
+// Only returns false if pec checks fails
+bool AP_BattMonitor_SMBus_Generic::initialize(uint32_t tnow)
 {
-    // get length
-    uint8_t bufflen;
-    // read byte (first byte indicates the number of bytes in the block)
-    if (!_dev->read_registers(reg, &bufflen, 1)) {
-        return 0;
+    if(_initialized){
+        return true;
     }
 
-    // sanity check length returned by smbus
-    if (bufflen == 0 || bufflen > SMBUS_READ_BLOCK_MAXIMUM_TRANSFER) {
-        return 0;
+    // check if PEC is supported
+    if (!check_pec_support()) {
+        _initialized = true;
+        return false;
     }
 
-    // buffer to hold results (2 extra byte returned holding length and PEC)
-    const uint8_t read_size = bufflen + 1 + (_pec_supported ? 1 : 0);
-    uint8_t buff[read_size];
+    // check cell count
+    if (_state.healthy) {
+        // when battery first becomes healthy, start check of cell count
+        if (_cell_count_check_start_us == 0) {
+            _cell_count_check_start_us = tnow;
+        }
+        if (tnow - _cell_count_check_start_us > (SMBUS_CELL_COUNT_CHECK_TIMEOUT * 1e6)) {
+            // give up checking cell count after 15sec of continuous healthy battery reads
+            _cell_count_fixed = true;
+
+            // If smart battery logging or mavlink messages were waiting for cell_count_fixed send them
+            if (_log_smart_batt_info) {
+                Log_Write_BATI(_instance);
+                _log_smart_batt_info = false;
+            }
 
-    // read bytes
-    if (!_dev->read_registers(reg, buff, read_size)) {
-        return 0;
-    }
+            if (_mavlink_smart_batt_info) {
+                send_mavlink_smart_battery_info(_instance, _chan);
```

Comment:
You can't send here. This is the I2C timer thread, and the GCS_MAVLink stuff isn't thread safe. You'd either have to set a flag for the GCS to poll at a regular interval, or just fail the first request if the data isn't available, and wait for a second one (which would be my personal preference, as it removes 2 members, and simplifies the GCS interaction, as well as fixing interactions when multiple links request the data at once).

Reply:
Removed this so that we now just notify the GCS of failure to gather all of the data. One day I'll learn which threads can interact with each other. :)  I assume that what I was doing with the logging thread was ok then?